Run spiped only once with -1

gperciva opened the pull request:
I put the man-page documenting of `-1` in a separate commit since you might tell me it should be undocumented.  My vote (which has no value) is that we should still document it, because I won't be able to remember it otherwise, and it's not like anybody can do anything nasty if they know about `-1`.  (and even if they could, they could discover it from the source code anyway).

Since `spipe` can quit with `^D`, I didn't add `-1` there, but you might want me to do so anyway for consistency.

cperciva commented:
Err... I thought we were going to make `-1` handling just a single *connection*?  Which may involve many calls to `events_run`...

cperciva commented:
(And spipe doesn't need this since it already only handles one connection.)

gperciva commented:
Updated according to our discussion this afternoon.  If this is accepted, I'll revise spipe along the same lines for consistency.

cperciva commented on spiped/spiped.1:
```diff
@@ -115,5 +116,8 @@ Disable target address re-resolution.
 Print version number.
+.TP
+.B \-1
+Only accept one event, then exit.  Used for internal testing.
```
Needs to be updated.  "Exit after a connection closes." ?

cperciva commented on spiped/main.c:
```diff
@@ -285,20 +292,28 @@ main(int argc, char * argv[])
+	}
+
+	/* Clean up. */
+	dispatch_shutdown(dispatch_cookie);
```
In non-error-handling paths I usually have comments explaining what the things are which we're freeing.  e.g., "Stop accepting connections and shut down the dispatcher", "Shut down the events system", "Free arrays of resolved addresses", etc.

cperciva commented on spiped/dispatch.c:
```diff
@@ -98,6 +99,10 @@ callback_conndied(void * cookie)
 	/* We've lost a connection. */
 	A->nconn -= 1;
 
+	/* If only one connection was desired, quit event loop. */
```
POV: The dispatcher doesn't know anything about users or command lines or event loops; all it knows is what the call to dispatch_accept told it, i.e., that sometimes we're supposed to set this to a non-zero value when a connection closes.

cperciva commented:
Are you keeping test scripts somewhere?  Maybe worth adding them to the tree?

Also, just a reminder: You need a test script which creates two connections through `spiped -1` and then closes one.  Right now this will result in the second connection not being cleaned up.

gperciva commented:
I'll add test scripts in a later PR; I've it written down on my TODO list now.

I know (or suspected, at least) that spiped would be leaking some stuff; my goal in this PR was to get spiped into a state that I could easily test for those, then do memory cleanup in a separate PR.

gperciva commented:
Updated.  As per our offline discussion, I've left my updates as separate commits; if they look ok, I'll rebase the patch so that the history looks nicer.

cperciva commented on spiped/main.c:
```diff
@@ -285,20 +292,36 @@ main(int argc, char * argv[])
 		goto err5;
 	}
 
 	/* Infinite loop handling events. */
```
`/* Loop until an error occurs, or a connection closes if running with -1. */`

cperciva commented:
Looks good aside from one comment tweak.  Clean up the commit history and then I'll merge.

gperciva commented:
Updated with comment change and cleaned up history.  The proposed comment doesn't fit onto one line, so I expanded it slightly to make use of the extra space:

```
/*
 * Loop until an error occurs, or a connection closes if the
 * command-line argument -1 was given.
 */
```
